control-plane/connect-inject/controllers/pod/pod_controller.go

```diff
@@ -432,10 +445,266 @@ func (r *Controller) writeHealthStatus(ctx context.Context, pod corev1.Pod) erro
+			}
+			if strings.TrimSpace(pieces[1]) == "port" {
+				portName = strings.TrimSpace(pieces[0])
+			} else {
+				return &pbmesh.Upstream{}, fmt.Errorf("upstream structured incorrectly: %s", rawUpstream)
+			}
+		default:
+			return &pbmesh.Upstream{}, fmt.Errorf("upstream structured incorrectly: %s", rawUpstream)
+		}
+	}
+
+	var upstream pbmesh.Upstream
+	if port > 0 {
+		upstream = pbmesh.Upstream{
+			DestinationRef: &pbresource.Reference{
+				Tenancy: &pbresource.Tenancy{
+					Partition: partition,
+					Namespace: namespace,
+					PeerName:  peer,
+				},
+				Name: svcName,
+			},
+			DestinationPort: portName,
+			Datacenter:      datacenter,
+			ListenAddr: &pbmesh.Upstream_IpPort{
+				IpPort: &pbmesh.IPPortAddress{
+					Port: uint32(port),
+				},
+			},
+		}
+	}
+
+	return &upstream, nil
+}
 
-//func (r *Controller) deleteUpstreams(pod corev1.Pod) error
+// processUnlabeledUpstream processes an upstream in the format:
+// [service-port-name].[service-name].[service-namespace].[service-partition]:[port]:[optional datacenter].
+// There is no unlabeled field for peering.
+// TODO: enable dc and peer support when ready, currently return errors if set. We also most likely won't need to return an error at all.
+func (r *Controller) processUnlabeledUpstream(pod corev1.Pod, rawUpstream string) (*pbmesh.Upstream, error) {
+	var portName, datacenter, svcName, namespace, partition, peer string
+	var port int32
+	var upstream pbmesh.Upstream
+
+	parts := strings.SplitN(rawUpstream, ":", 3)
+
+	port, _ = common.PortValue(pod, strings.TrimSpace(parts[1]))
+
+	// If Consul Namespaces or Admin Partitions are enabled, attempt to parse the
+	// upstream for a namespace.
+	if r.EnableConsulNamespaces || r.EnableConsulPartitions {
+		pieces := strings.SplitN(parts[0], ".", 4)
+		switch len(pieces) {
+		case 4:
+			partition = strings.TrimSpace(pieces[3])
+			fallthrough
+		case 3:
+			namespace = strings.TrimSpace(pieces[2])
+			fallthrough
+		default:
+			svcName = strings.TrimSpace(pieces[1])
+			portName = strings.TrimSpace(pieces[0])
+		}
+	} else {
+		pieces := strings.SplitN(parts[0], ".", 2)
+		svcName = strings.TrimSpace(pieces[1])
+		portName = strings.TrimSpace(pieces[0])
+	}
+
+	// parse the optional datacenter
+	if len(parts) > 2 {
+		// TODO: uncomment and remove error when datacenters supported
+		//datacenter = strings.TrimSpace(parts[2])
+		return &pbmesh.Upstream{}, fmt.Errorf("upstream currently does not support datacenters: %s", rawUpstream)
+	}
+
+	if port > 0 {
+		upstream = pbmesh.Upstream{
+			DestinationRef: &pbresource.Reference{
+				Tenancy: &pbresource.Tenancy{
+					Partition: partition,
+					Namespace: namespace,
+					PeerName:  peer,
+				},
+				Name: svcName,
+			},
+			DestinationPort: portName,
+			Datacenter:      datacenter,
+			ListenAddr: &pbmesh.Upstream_IpPort{
+				IpPort: &pbmesh.IPPortAddress{
+					Port: uint32(port),
+				},
+			},
+		}
+	}
+	return &upstream, nil
+}
```

Comment:
Same comments here as the labeled-style:
1. need explicit tenenacy
2. need Type in DestinationRef
3. need IP in the listener